ps3.py

```diff
@@ -0,0 +1,361 @@
+      from the hand.
+
+    * An invalid word is rejected, and a message is displayed asking
+      the user to choose another word.
+
+    * After every valid word: the score for that word is displayed,
+      the remaining letters in the hand are displayed, and the user
+      is asked to input another word.
+
+    * The sum of the word scores is displayed when the hand finishes.
+
+    * The hand finishes when there are no more unused letters.
+      The user can also finish playing the hand by inputing two
+      exclamation points (the string '!!') instead of a word.
+
+      hand: dictionary (string -> int)
+      word_list: list of lowercase strings
+      returns: the total score for the hand
+
+    """
+    # Keep track of the total score
+    # As long as there are still letters left in the hand:
+    total_score = 0
+    while True:
+        if calculate_handlen(hand) == 0:
+            print("Ran out of letters")
+            break
+        # Display the hand
+        display_hand(hand)
+        # Ask user for input
+        word = input("Enter word, or “!!” to indicate that you are finished: ")
+        if word == STOP_GAME_SYMBOL:
+            break
+        if is_valid_word(word, hand, word_list):
+            # If the word is valid:
+            # Tell the user how many points the word earned,
+            # and the updated total score
+            current_score = get_word_score(word, HAND_SIZE)
+            total_score += current_score
+            print(f'"{word}" earned {current_score} points. Total: {total_score} points')
+        else:
+            # Otherwise (the word is not valid):
+            # Reject invalid word (print a message)
+            print("This is not a valid word. Please choose another word.")
+        hand = update_hand(hand, word)
+        print()
+    # Return the total score as result of function
+    print("Total score for this hand:", total_score)
+    return total_score
+
+
+def substitute_hand(hand, letter):
+    """
+    Allow the user to replace all copies of one letter in the hand (chosen by user)
+    with a new letter chosen from the VOWELS and CONSONANTS at random. The new letter
+    should be different from user's choice, and should not be any of the letters
+    already in the hand.
+
+    If user provide a letter not in the hand, the hand should be the same.
+
+    Has no side effects: does not mutate hand.
+
+    For example:
+        substitute_hand({'h':1, 'e':1, 'l':2, 'o':1}, 'l')
+    might return:
+        {'h':1, 'e':1, 'o':1, 'x':2} -> if the new letter is 'x'
+    The new letter should not be 'h', 'e', 'l', or 'o' since those letters were
+    already in the hand.
+
+    hand: dictionary (string -> int)
+    letter: string
+    returns: dictionary (string -> int)
+    """
+    hand_modified = hand.copy()
+    letters_all = string.ascii_lowercase
+    while True:
+        randomized_letter = random.choice(letters_all)
```

Comment:
You could just remove all letters from hand in letters_all, then you wouldn't need this loop

Reply:
done